test/serdes_test.c

```diff
@@ -0,0 +1,194 @@
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include <openssl/evp.h>
+#include <openssl/pem.h>
+#include <openssl/rsa.h>
+#include <openssl/x509.h>
+#include <openssl/asn1.h>
+#include <openssl/serializer.h>
+#include <openssl/deserializer.h>
+
+#include "testutil.h"
+
+static EVP_PKEY *key_RSA = NULL;
+
+static EVP_PKEY *make_RSA(int rsa_type)
+{
+    EVP_PKEY *pkey = NULL;
+    EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new_id(rsa_type, NULL);
+
+    /*
+     * No real need to check the errors other than for the cascade
+     * effect.  |pkey| will imply remain NULL if something goes wrong.
+     */
+    (void)(ctx != NULL
+           && EVP_PKEY_keygen_init(ctx) > 0
+           && EVP_PKEY_CTX_set_rsa_keygen_bits(ctx, 2048) > 0
+           && EVP_PKEY_keygen(ctx, &pkey) > 0);
+    EVP_PKEY_CTX_free(ctx);
+
+    return pkey;
+}
+
+typedef int (checker)(int type, const void *data, size_t data_len);
+typedef void (dumper)(const char *label, const void *data, size_t data_len);
+
+static int test_serialize_deserialize(EVP_PKEY *pkey,
+                                      const char *pass, const char *pcipher,
+                                      checker *check_cb, dumper *dump_cb,
+                                      const char *ser_propq)
+{
+    BIO *mem_ser = NULL, *mem_deser = NULL;
+    OSSL_SERIALIZER_CTX *sctx = NULL;
+    void *serialized = NULL;
+    long serialized_len;
+    EVP_PKEY *pkey2 = NULL;
+    OSSL_DESERIALIZER_CTX *dctx = NULL;
+    const unsigned char *upass = (const unsigned char *)pass;
+    int ok = 0;
+
+    if (!TEST_ptr(sctx = OSSL_SERIALIZER_CTX_new_by_EVP_PKEY(pkey, ser_propq))
+        || (pass != NULL
+            && !OSSL_SERIALIZER_CTX_set_passphrase(sctx, upass, strlen(pass)))
+        || (pcipher != NULL
+            && !OSSL_SERIALIZER_CTX_set_cipher(sctx, pcipher, NULL))
+        || !TEST_ptr(mem_ser = BIO_new(BIO_s_mem()))
+        || !TEST_true(OSSL_SERIALIZER_to_bio(sctx, mem_ser))
+        || !TEST_long_gt(serialized_len =
+                         BIO_get_mem_data(mem_ser, &serialized), 0))
+        goto end;
+
+    if (!check_cb(EVP_PKEY_base_id(pkey), serialized, serialized_len))
+        goto end;
+
+    if (!TEST_ptr(dctx = OSSL_DESERIALIZER_CTX_new_by_EVP_PKEY(&pkey2, NULL,
+                                                               NULL))
+        || (pass != NULL
+            && !OSSL_DESERIALIZER_CTX_set_passphrase(dctx, upass,
+                                                     strlen(pass)))
+        || (pcipher != NULL
+            && !OSSL_DESERIALIZER_CTX_set_cipher(dctx, pcipher, NULL))
+        || !TEST_ptr(mem_deser = BIO_new_mem_buf(serialized, serialized_len))
+        || !TEST_true(OSSL_DESERIALIZER_from_bio(dctx, mem_deser))
+        || !TEST_int_eq(EVP_PKEY_cmp(pkey, pkey2), 1))
```

Comment:
I know you are doing a EVP_PKEY_cmp here... But it may be worthwhile making the serialize bit a function and calling it again on pkey2 to make sure the serialized result is the same.

Reply:
Good idea!

Reply:
Done